File: extensions/pyre/memory/maps.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2025 all rights reserved


// the module initializers
#include "__init__.h"
// my maps
#include "maps.h"


// wrappers over {pyre::memory::map_t} template expansions
void
pyre::py::memory::maps::__init__(py::module & memory)
{
    // create a submodule for the map stores
    auto maps = memory.def_submodule(
        // the name
        "maps",
        // its docstring
        "file backed storage");

    // get the pile of map types
    using maps_t = pyre::memory::maps_t;
    // build the classes
    expand(maps, maps_t {});

    // all done
    return;
}


// end of file

File: tests/pyre.ext/memory/maps.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2025 all rights reserved


# support
import itertools


# the driver
def maps():
    """
    Verify that the map expansions are all present
    """
    # get the module
    from pyre.extensions.pyre.memory import maps

    # given access rights
    permissions = ["", "Const"]
    # and basic types
    types = [
        # ints
        "Int8",
        "Int16",
        "Int32",
        "Int64",
        "UInt8",
        "UInt16",
        "UInt32",
        "UInt64",
        # floats
        "Float",
        "Double",
        # complex
        "ComplexFloat",
        "ComplexDouble",
    ]
    # build all possible names
    names = [
        f"Map{type}{permission}"
        for permission, type in itertools.product(permissions, types)
    ]
    # go through them
    for name in names:
        # and verify they are all present
        assert getattr(maps, name)

    # all done
    return


# main
if __name__ == "__main__":
    maps()


# end of file
